File: framework/languages/dotnet/e2e/DiscoveryRegistryHa/Server/Consumer/ConsumerHostFactory.cs
using DiscoveryRegistryHa.Shared;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using Zlink.Framework.AspNetCore;
using Zlink.Framework.Contracts.Channels;
using Zlink.Framework.Contracts.Dispatch;

namespace DiscoveryRegistryHa.Server.Consumer;

internal static class ConsumerHostFactory
{
    public static WebApplication Create(string[] args)
    {
        var options = ConsumerOptions.Parse(args);
        Directory.CreateDirectory(options.LogDir);

        var builder = WebApplication.CreateBuilder(args);
        builder.Logging.ClearProviders();
        builder.Logging.AddSimpleConsole(console =>
        {
            console.SingleLine = true;
            console.TimestampFormat = "HH:mm:ss.fff ";
        });
        builder.WebHost.UseUrls(options.HttpUrl);
        builder.Services.AddZLinkFramework(framework =>
        {
            foreach (var endpoint in options.DiscoveryEndpoints)
            {
                framework.UseDiscovery().AddRegistryEndpoint(endpoint);
            }

            framework.ConfigureDispatch()
                .MessageFlow(ZLinkMessageFlowLogMode.KeyTransitions)
                .TraceLogFile(Path.Combine(options.LogDir, $"{options.Rid}-flow.log"))
                .TraceLabel(options.Rid);
            framework.AddClientServerChannel(DiscoveryRegistryHaNames.Channel)
                .EnableClient();
        });

        var app = builder.Build();
        app.MapGet("/health", () => Results.Ok(new { status = "ready", options.Rid }));
        app.MapPost("/profile/request", async (ProfileRequest request, IZLinkChannelClient client) =>
        {
            var reply = await client.RequestToChannel(DiscoveryRegistryHaNames.Channel, request)
                .PacketName("ProfileRequest")
                .Timeout(TimeSpan.FromSeconds(3))
                .Async<ProfileReply>();
            return Results.Ok(reply);
        });
        app.MapPost("/profile/request/wait", async (
            ProfileRequest request,
            IZLinkChannelClient client,
            CancellationToken cancellationToken) =>
        {
            var deadline = DateTimeOffset.UtcNow + TimeSpan.FromSeconds(10);
            Exception? last = null;
            while (DateTimeOffset.UtcNow < deadline)
            {
                try
                {
                    var reply = await client.RequestToChannel(DiscoveryRegistryHaNames.Channel, request)
                        .PacketName("ProfileRequest")
                        .Timeout(TimeSpan.FromSeconds(3))
                        .Async<ProfileReply>();
                    return Results.Ok(reply);
                }
                catch (Exception ex)
                {
                    last = ex;
                    await Task.Delay(TimeSpan.FromMilliseconds(100), cancellationToken);
                }
            }

            throw new TimeoutException("Timed out waiting for profile request routing.", last);
        });
        app.MapPost("/shutdown", (IHostApplicationLifetime lifetime) =>
        {
            lifetime.StopApplication();
            return Results.Ok(new { status = "stopping" });
        });
        return app;
    }
}

File: framework/languages/node/e2e/DiscoveryRegistryHa/Server/Consumer/consumer-host-factory.ts
import fs from 'node:fs';
import { Module } from '@nestjs/common';
import { NestFactory } from '@nestjs/core';
import { ZLinkMessageFlowLogMode, type ZLinkLocationRuntimeQuery, type ZLinkChannelClient } from '@zlink-systems/framework';
import { ZLINK_CHANNEL_CLIENT, ZLINK_LOCATION_RUNTIME_QUERY, ZLinkModule, zlinkFramework } from '@zlink-systems/nestjs';
import { ChannelNames } from '../../Shared/messages';
import { createRedisLocationStore, storeFailureLocationOptions } from '../../Shared/location-store';
import { validateConsumerOptions } from './Configuration/consumer-options';
import type { ConsumerOptions } from './Configuration/consumer-options';
import { DISCOVERY_OPTIONS, createDiscoveryConfigurationModule } from '../../configuration';
import { createConsumerEndpoints } from './Endpoints/consumer-endpoints';
import { closeHttpServer, startHttpServer } from './Support/http-server';

export async function startConsumerHost(): Promise<void> {
  let stopping = false;

  const ConsumerModule = createConsumerModule();
  const app = await NestFactory.createApplicationContext(ConsumerModule, { logger: false, abortOnError: false });
  const options = app.get(DISCOVERY_OPTIONS, { strict: false }) as ConsumerOptions;
  const channel = app.get(ZLINK_CHANNEL_CLIENT, { strict: false }) as ZLinkChannelClient;
  const locationQuery = app.get(ZLINK_LOCATION_RUNTIME_QUERY, { strict: false }) as ZLinkLocationRuntimeQuery;
  const server = await startHttpServer(options.httpUrl, createConsumerEndpoints(channel, locationQuery, () => { stopping = true; }));
  while (!stopping) {
    await new Promise((resolve) => setTimeout(resolve, 100));
  }
  await closeHttpServer(server);
  await app.close();
}

function createConsumerModule(): Function {
  class ConsumerModule {}
  const configuration = createDiscoveryConfigurationModule(validateConsumerOptions);
  Module({
    imports: [
      configuration,
      ZLinkModule.forRootFactory({
        imports: [configuration], inject: [DISCOVERY_OPTIONS],
        useFactory: (value: unknown) => {
          const options = value as ConsumerOptions;
          fs.mkdirSync(options.logDir, { recursive: true });
          const builder = zlinkFramework();
          builder
            .configureDispatch()
              .messageFlow(ZLinkMessageFlowLogMode.KeyTransitions)
              .traceLogFile(`${options.logDir}/${options.traceLabel}-flow.log`)
              .traceLabel(options.traceLabel);
          builder.addLocationStore(createRedisLocationStore(options));
          Object.assign(builder.configureLocations(), storeFailureLocationOptions());
          builder.addRouteMesh(ChannelNames.profile).peerConnections();
          return builder.build();
        }
      })
    ]
  })(ConsumerModule);
  return ConsumerModule;
}
